### PR TITLE
Add additional open container api to HumanEntity

### PR DESCRIPTION
**Pull request by JRoy**
another round of "i want this in essentials but i don't want to enter nms hell"

### CODE REVIEW

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

**Review by kashike — COMMENTED** [review]
It feels like a lot of the implementation duplicates code - you should be able to reduce it.

File: Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch
Comment (Proximyst) [inline Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch]: Wouldn't it be easier for users to just specify inventory type?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: Just set to `!force`

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: Async catcher; this modifies state.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch
Comment (JRoy) [reply Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch]: Existing API to do this doesn't work correctly for these use cases (UnsupportedOperationExceptions + improper reachability checks). Due to the way containers are written, speific casting is required in order to properly so one general purpose method would end up being as long as indivial methods.

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch
Comment (JRoy) [reply Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: I used code from the existing `openEnchant` method as a reference. I didn't notice this and it appears that this is a regression in the craftbukkit implementation.

Will fix this in these new methods

**Comment by JRoy** [comment]
@kashike Code duplication has been reduced as much as possible.

Due to the way these blocks are written, there is a lot of manual casting needed so that is why I had to shove a switch case in the helper method. Let me know if you want me to do that in a different way.

**Review by Proximyst — DISMISSED** [review]
I'll merge tomorrow if no-one else beats me to it.

**Review by aikar — DISMISSED** [review]

**Comment by MiniDigger** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
thanks!

**Pull request merged**
